aten/src/ATen/native/TensorShape.cpp

```diff
@@ -563,6 +563,82 @@ static void fastCatOutDim0(const Tensor& out, const MaterializedITensorListRef&
   TORCH_CHECK(outBytes == totalBytes);
 }
 
+static void parallelFastCatOutDim0(
+    const Tensor& out,
+    const MaterializedITensorListRef& inputs) {
+  // Partition inputs into parallel chunks based on number of elements
+  size_t out_elements = out.numel();
+  if (out_elements < at::internal::GRAIN_SIZE || at::get_num_threads() == 1) {
+    fastCatOutDim0(out, inputs);
+    return;
+  }
+  size_t total_elements = 0;
+  size_t copy_elements = 0;
+  std::vector<std::pair<size_t, size_t>> input_chunks;
+  size_t start = 0;
+  size_t end = 0;
+  for (size_t i = 0; i < inputs.size(); i++) {
+    const Tensor& input = inputs.at(i);
+    const size_t input_elements = input.numel();
+    total_elements += input_elements;
+    TORCH_CHECK(out_elements >= total_elements);
+    if (i == inputs.size() - 1) {
+      TORCH_CHECK(out_elements == total_elements);
+    }
```

Comment:
nit: why do we need to waste cycles doing these checks inside the loop instead of right after it?